metawatch.js

```diff
@@ -3,26 +3,66 @@
 const fs = require('fs');
 const path = require('path');
 
-const watch = (targetPath, listener) => {
-  fs.readdir(targetPath, { withFileTypes: true }, (err, files) => {
-    for (const file of files) {
-      if (file.isDirectory()) {
-        const dirPath = path.join(targetPath, file.name);
-        watch(dirPath, listener);
-      }
-    }
-    fs.watch(targetPath, (event, fileName) => {
-      const filePath = path.join(targetPath, fileName);
-      try {
-        fs.stat(filePath, (err, stats) => {
-          if (stats.isDirectory()) watch(filePath, listener);
-        });
-      } catch {
-        return;
+const DEFAULT_DEBOUNCE = 10;
+
+class Watcher {
+  constructor(config = {}) {
+    this.debounce = config.debounce || DEFAULT_DEBOUNCE;
+
+    this.watchers = [];
+    this.isClosed = false;
+  }
+
+  watch(targetPath, listener) {
+    const wrappedListener = this.debounceWrapper(listener, this.debounce);
+
+    fs.readdir(targetPath, { withFileTypes: true }, (err, files) => {
+      if (this.isClosed) return;
+
+      for (const file of files) {
+        if (file.isDirectory()) {
+          const dirPath = path.join(targetPath, file.name);
+          this.watch(dirPath, listener);
+        }
       }
-      listener(event, fileName);
+      const watcher = fs.watch(targetPath, (event, fileName) => {
+        const filePath = path.join(targetPath, fileName);
+        try {
+          fs.stat(filePath, (err, stats) => {
+            if (stats.isDirectory()) {
+              this.watch(filePath, listener);
```

Comment:
Need to check this folder, maybe it is already in `this.watchers`

Reply:
Changed the logic of storing `watchers`. And tested all possible manipulations with watched folders, duplication of watchers is not happaning. I think, we do not need additional check for it, because it will restrict such case of usage:
```
  const watcher = new Watcher();
  watcher.watch('./test', () => console.log('watch1'));
  watcher.watch('./test', () => console.log('watch2'));